cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -659,8 +659,24 @@ func (s *service) Stats(ctx context.Context, r *taskAPI.StatsRequest) (*taskAPI.
 	}, nil
 }
 
+func (s *service) killInitChildren(e runcC.Exit) {
+	for _, cp := range s.running[e.Pid] {
+		// Ensure all children are killed
+		if ip, init := cp.Process.(*process.Init); init {
+			if runc.ShouldKillAllOnExit(s.context, cp.Container.Bundle) {
+				log.G(s.context).Infof("kill init's (pid %d) children in container %s", e.Pid, cp.Container.ID)
+				if err := ip.KillAll(s.context); err != nil {
+					log.G(s.context).WithError(err).WithField("id", ip.ID()).
+						Error("failed to kill init's children")
+				}
+			}
+		}
+	}
+}
+
 func (s *service) processExits() {
 	for e := range s.ec {
+		s.killInitChildren(e)
```

Comment:
I don't believe this change alone fixes the issue. Even if we kill the execs when the init exits, what's there to prevent new execs from being consistently started (and thereby incrementing `s.pendingExecs[c]`, never allowing us to process the init's exit?

Reply:
```
runc             1706700 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process3195875960 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/0543f08f54 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706694 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process2680626484 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/0e0344ccb6 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706701 1701666   0 
runc             1706723 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process250304382 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/9877926623 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706721 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process2161178712 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/fc7c8acdb1 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706727 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process662906153 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/94676ded44 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706722 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process3020290985 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/1dd5bab6be e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706732 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process3697317412 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/3b8665e38d e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706763 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process2925547936 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/ab4e645ecd e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706814 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process1989814169 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/f0d1ada99d e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706819 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process417503049 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/bbbae26a56 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1706823 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json exec --process /tmp/runc-process94538854 --detach --pid-file /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/9c2a5305e1 e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
runc             1707176 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json kill --all e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08 9
runc             1708857 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json kill --all e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08 9
runc             1708864 1701666   0 /usr/bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json delete e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
containerd-shim  1708878 1700679   0 /bin/containerd-shim-runc-v2 -namespace k8s.io -address /run/containerd/containerd.sock -publish-binary /usr/bin/containerd -id e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08 -bundle /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08 delete
runc             1708889 1708878   0 /bin/runc --root /run/containerd/runc/k8s.io --log /run/containerd/io.containerd.runtime.v2.task/k8s.io/e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08/log.json --log-format json delete --force e9371f5ed835ba23c967541c4777265166b42a542e280351f1997215f98a4e08
```
I catch runc exec use  /usr/share/bcc/tools/execsnoop | grep --color=auto runc
 after runc kill will not generate any runc exec.

Reply:
how about change  handleProcessExit to sendProcessExit? we just need  init exited is last sent. we should call p.SetExited(e.Status) as soon as possible